core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -310,15 +305,33 @@ public void registerFunctionScoreParser(ScoreFunctionParser<? extends ScoreFunct
      *        registered. So it is the name that the query should use as its {@link NamedWriteable#getWriteableName()}.
      */
     public <QB extends QueryBuilder<QB>> void registerQuery(Writeable.Reader<QB> reader, QueryParser<QB> parser, String... names) {
```

Comment:
This doesn't directly concern changes in this PR, but the first name should be mandatory here. What about making this signature `String queryName, String alternativeNames` or something along those lines?